python/ray/data/datasource/bigquery_datasource.py

```diff
@@ -0,0 +1,199 @@
+            dataset_id = destination.split(".")[-2]
+            table_id = destination.split(".")[-1]
+        else:
+            self._validate_dataset_table_exist(self._project_id, self._dataset)
+            dataset_id = self._dataset.split(".")[0]
+            table_id = self._dataset.split(".")[1]
+
+        bqs_client = bigquery_storage.BigQueryReadClient()
+        table = f"projects/{self._project_id}/datasets/{dataset_id}/tables/{table_id}"
+
+        if parallelism == -1:
+            parallelism = None
+        requested_session = types.ReadSession(
+            table=table,
+            data_format=types.DataFormat.ARROW,
+        )
+        read_session = bqs_client.create_read_session(
+            parent=f"projects/{self._project_id}",
+            read_session=requested_session,
+            max_stream_count=parallelism,
+        )
+
+        read_tasks = []
+        print("Created streams:", len(read_session.streams))
+        if len(read_session.streams) < parallelism:
+            print(
+                "The number of streams created by the "
+                + "BigQuery Storage Read API is less than the requested "
+                + "parallelism due to the size of the dataset."
+            )
+
+        for stream in read_session.streams:
+            # Create a metadata block object to store schema, etc.
+            metadata = BlockMetadata(
+                num_rows=None,
+                size_bytes=None,
+                schema=None,
+                input_files=None,
+                exec_stats=None,
+            )
+
+            # Create a no-arg wrapper read function which returns a block
+            read_single_partition = lambda stream=stream, kwargs=self._kwargs: [
+                _read_single_partition(stream, kwargs)
+            ]
+
+            # Create the read task and pass the wrapper and metadata in
+            read_task = ReadTask(read_single_partition, metadata)
+            read_tasks.append(read_task)
+
+        return read_tasks
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return None
+
+    def _validate_dataset_table_exist(self, project_id: str, dataset: str) -> None:
+        client = bigquery.Client(project=project_id)
+        dataset_id = dataset.split(".")[0]
+        try:
+            client.get_dataset(dataset_id)
+        except exceptions.NotFound:
+            raise ValueError(
+                "Dataset {} is not found. Please ensure that it exists.".format(
+                    dataset_id
+                )
+            )
+
+        try:
+            client.get_table(dataset)
+        except exceptions.NotFound:
+            raise ValueError(
+                "Table {} is not found. Please ensure that it exists.".format(dataset)
+            )
+
+
+class BigQueryDatasource(Datasource):
+    def create_reader(self, **kwargs) -> Reader:
+        return _BigQueryDatasourceReader(**kwargs)
+
+    def do_write(
```

Comment:
do_write() is deprecated, could you port this to use the new write() interface?

Reply:
Sure. How would we reference the metadata blocks since the new write() interface does not list metadata as an argument provided?

Reply:
I've also noticed that there is no `cached_remote_fn` call to write each block (at least, how MongoDB datasource was updated for the Write API). Write does not return a list of write tasks anymore as well. How are writes supposed to be parallelized?

Reply:
In the new write API, datasets itself creates the tasks used for write, and calls datasource.write() in remote tasks for the blocks assigned to each write task (this allows optimizations such as fusing write tasks with previous stages).

> Sure. How would we reference the metadata blocks since the new write() interface does not list metadata as an argument provided?

You could use BlockAccessor.for_block(block).get_metadata() to compute metadata for the block.

Reply:
Got it, I've upgraded to use the write() interface. I'm a bit less familiar with how it works than do_write so let me know if anything seems off.